### DISCUSSION TITLE
Clangd parsed the c code incorrectly

### DISCUSSION THREAD

**Discussion by han-shao-jun** [clangd/clangd#1395] [Q&A]
Hi！I met with difficulties in my project. I would be grateful if someone could answer my hunger question.
## Problem description
My project is edited with vscode and managed with cmake. The following is my CMakeLists.txt file, which adds a predefined macro called debug. 
```cmake
cmake_minimum_required(VERSION 3.22)
project(test C)

set(CMAKE_C_STANDARD 99)
add_compile_definitions(debug)

add_executable(test main.c)

```
This is my compile_ commands. Json file. 
```json
[
{
  "directory": "D:/code/c/test/build",
  "command": "D:\\Programs\\msys64\\clang64\\bin\\clang.exe -Ddebug -g -std=gnu99 -o CMakeFiles\\test.dir\\main.c.obj -c D:\\code\\c\\test\\main.c",
  "file": "D:\\code\\c\\test\\main.c"
}
]
```
Using the following code, clangd will report incorrect syntax, but in fact, there is no syntax error, because I can compile and run.
```c
#include <stdio.h>

#undef debug
int a = 0;

#ifndef debug
int i = 0;
#endif

int main()
{
    printf("i = %d", i);
    return 0;
}


```

## System information
Output of clangd --version
```shell
$ clangd --version
clangd version 15.0.3
Features: windows
Platform: x86_64-w64-windows-gnu
```
OS: windows11

**Comment by HighCommander4** [comment]
This is a clangd bug, related to the preamble optimization (see [this comment](https://github.com/clangd/clangd/issues/200#issuecomment-554973739) for some further discussion).

A workaround is to insert any declaration before the `#undef` (for example `int dummy;`) to cause the `#undef` to be parsed as part of the main-file and not the preamble.

**Reply by HighCommander4** [reply]
Since I do believe it's a legitimate bug and should ideally be fixed, I filed https://github.com/clangd/clangd/issues/1396 for it.

**Reply by han-shao-jun** [reply]
![image](https://user-images.githubusercontent.com/62562088/204239280-fc7c22b7-27a2-4c5c-bfab-8348b6a73678.png)
As shown in the figure above, clangd will report an error by mistake, but if I change to the figure below, clangd will not report an error. I guess this is a bug of clangd
![image](https://user-images.githubusercontent.com/62562088/204239352-4b81db65-1616-4bcc-989d-502fdf68d9e5.png)

**Reply by HighCommander4** [reply]
I believe the error disappearing when the `#ifndef debug`  follows the `#undef debug` with no declarations in between, is a symptom of the same bug (now the `#ifndef` is part of the preamble too).

**Reply by han-shao-jun** [reply]
Will this bug be fixed？

**Reply by HighCommander4** [reply]
@han-shao-jun I don't plan to work on it myself. Beyond that, I can't say. My sense is issues related to the preamble optimization tend to be tricky to fix.
